### ISSUE TITLE
Balancing outfit and cargo space

### ISSUE THREAD

**Issue by Architect-Luke** [endless-sky/endless-sky#2184]
It is my belief that we have a fundamental problem in the ship customization system. The conflicting points are thus:
1) there is a desire to make ships as customizable as possible without breaking things
2) there is a desire for tiered game progression

Outfits can be put into any ship that has the space, which means that the primary limiting factor for ship power is outfit space.

I propose removing mass expansions. I don't want to write an essay on the benefits, but please don't dismiss it out of hand. A ship's maximum outfit space will be static, which means its power will be effectively capped without the need to consider every ton of cargo space.

This is the reason why cargo space is so limited. Outfit space is king, and every 1 ton of cargo is 0.75 tons of outfit space. That means that every 80 tons of cargo is a Nanobattery. Every 122 tons of cargo space is a Large Systems Core.

I think the impact will be only positive. We will see MUCH greater ship variety and options for players. Alternatively, I propose a hard cap for mass expansions per ship.

**Comment by Architect-Luke** [comment]
Note that I'm suggesting removing what has basically been one of my favorite outfits. I have an emotional investment in the mass expansion, as I'm sure many of you do. However, I truly believe it's for the good of the game.

**Comment by SonEasterZombie** [comment]
Don't forget that the option to expand mass was put in to create greater change in ships. It allows the reverse to happen too, with Bactrians that have more cargo than pretty much any cargo ship and things like that. Not only this but I also feel that the ability to exchange cargo space for outfit space can tell a story too, like how all the Kor Ships have outfit space expansions to show that they don't care about trade and their ships are all modified for warfare.

While removing the expansions could create a need for the player to purchase a larger variety of ships and could potentially nerf ships such as the Bactrian (Something that needs to be done)   1) I'm not even sure if M.Z has massive specialized fleets in mind for this game and   2) It would get rid of a lot of the charm of certain ships and would require either major buffing of republic and other outfit expanded ships or a major nerfing of their capabilities in combat.

**Comment by KiLEdEnNis** [comment]
> which means that the primary limiting factor for ship power is outfit space.

and general stats: hull/shields (& stuff like weapon hardpoints, cargo space, bunks, but primarily the durability)

and the mass/cargo expansions have a 4:3 ratio which makes them not that OP, it adds some extra outfit or cargo space at the cost of the other

you still have to grind for a nanotech battery (& the Quarg won't turn friendly after ya plunder them once as of a new update), and a large systems core is hard to come by
it's the player's own choice to give up outfit/cargo space for the other one

basically, it let's the player think "do I want the ~ space or do I want to convert it", I don't see a reason to really remove them, they're only an issue on the Bactrian, they just add a bit of a choice for other ships and they enable retrofitted ships

as for hard caps, MZ didn't want to limit how much of something you could install

**Comment by endless-sky** [comment]
The idea that I've been playing with is to make mass expansions come with some sort of penalty. The explanation would be that rearranging the interior of your ship in such a fundamental way is detrimental to your ship's performance: decreasing heat dissipation, or increasing "drag," or even decreasing hull strength. The trouble is that if you get more outfit space you can always use some of it for cooling, so decreasing heat dissipation turns out not to be a very effective penalty. But making the ship slower and flimsier might work.

Say, for example, that each mass expansion decreases hull and shields by 250 and adds 0.5 to drag. That means a Bactrian with a full 26 mass expansions would have its top speed nearly cut in half and its shields / hull reduced from 17500 / 8600 to 11000 / 2100. It could still be usable as a flagship for boarding, but even if packed full of shield generators it would have to avoid pitched combat.

**Comment by KiLEdEnNis** [comment]
mass expansions could maybe also add some extra mass to make turning slower?

**Comment by Amazinite** [comment]
Mass expansions already add extra mass.

**Comment by Bladewood** [comment]
*cloaking device*

**Comment by Bladewood** [comment]
What if we tried to do something retarded, like installing a bunch of cargo expansions and then mass expansions?

**Comment by KiLEdEnNis** [comment]
@Amazinite I know, but maybe extra mass (although it'd "take" 5 space but add more than 5 mass, would be weird)

**Comment by KiLEdEnNis** [comment]
-100 shield or so ok, but don't even touch hull, since you can get instances where ya take off & die because your ship has no hull (or make it that a lot of expansions reduce nearly 50%, with an asymptote at the 50%, so every extra expansion "punishes" less)
although I feel like that makes it annoying and confusing as well

**Comment by Architect-Luke** [comment]
Attaching penalties is the wrong direction. It won't scale and it will make balance a nightmare. The point here is to make balance easier and allow more ship variety.

You can have ships with any stats and cargo space, if you don't have to worry about that cargo space turning into general purpose space. This means we can have larger ships that don't suck, without them being potentially overpowered.

**Comment by Architect-Luke** [comment]
Think of being able to have much more variety in your ships. Cargo capacity on heavy warships, freighters that can carry more significant amounts, freighters that don't have terrible outfit space, etc. You can have a Raider with actual cargo space for raiding, and equivalents at other tiers.

**Comment by MessyMix** [comment]
So are you just saying we buff everything and remove Mass Expansions?

Because I can retrofit ships to give me exactly what you've just said.

**Comment by retrue** [comment]
My humble opinion:
In any case, do not remove completely mass expansions. I don't like the idea of penalizing expansions either. And if you limit their number, there are several options:
1.- Specify the number of maximum mass expansions for each model of ship and put that information in the specs of each ship. For example, Bactrians, 10 mass expansions maximum.
2.- Specify a percentage for all ships. For example, any ship can only convert 50% of its cargo space in outfit space. Or 40%, whatever.
3.- Specify different percentages for each type of ship depending on their mass/size or their function. For example: warships can convert 40% of their cargo in outfit space, transports 60%. Or small ships 90% and big ships 30%.

**Comment by DingusShingleton** [comment]
How about this.... kind of outside the box but bear with me. The idea of mass/cargo expansions as an outfit always kind of rubbed me the wrong way. What is it? I'm buying units of some presumably tangible thing right? Or is it simply a game mechanic necessity? I don't know, but either way here's my thought:

Get rid of them. Instead, once the maximum outfit space is reached, allow outfits to be installed directly into the cargo bay. Have a first-run pop up that says "Hey! I see that you are installing outfits into your cargo bay! That's fine, but when you do that [penalty] will happen!". As it stands, they take up 33% more cargo than the listed outfit space, but additional penalties could easily be added. More mass, a percentage increase in cost (to make up for not having to buy expansions), additional heat/ton, or energy/ton, or really anything at all.

IMO this is a much more seamless and intuitive (not to mention realistic) way of dealing with it, and it also makes it easier to gradually penalize/limit the penalties of doing so without having hard caps on anything.

The only problem I can see is how the game will determine which outfits are installed in cargo. To simplify that issue away, it could just calculate penalties by the ton, regardless of outfit type. IE each ton of outfit loaded into cargo costs x credits, creates y heat, and uses z mass. Or whatever. Honestly I just came up with this so it's pretty half baked, but maybe it will inspire someone else to have an even better idea.

**Comment by jafdy** [comment]
What would you do about cargo expansions in this case?

**Comment by DingusShingleton** [comment]
Do the same thing? Pressing buy on a commodity when your cargo hold is full pops up a similar warning and then loads it into your available outfit space. I suppose the cost of doing so would have to be carefully balanced or else loading cargo into outfit space could easily become pointless.

**Comment by w0ntoan** [comment]
@DingusShingleton Cargo stats vs Outfit stats sounds like a logistical nightmare, deciding which area one should put items, and then managing that.

I'm with @KiLEdEnNis about not touching hull. Outfit space needs electrical wiring but not cargo, so all that unintended EM interference screws with shields? Okay. Large outfits aren't able to be balanced like many smaller cargo loads, so it causes engine response to resist imbalance? Okay.

Also, flat penalties would be easy to implement (it's already in there), but might be unfair for smaller ships?

**Comment by DingusShingleton** [comment]
Well the player would never have to decide. It would only load into the other area when the normal space was full. Or do you mean in terms of the code?

**Comment by Architect-Luke** [comment]
I know it might rub you the wrong way, but consider it more carefully. Right now we have large freighters with less cargo space than some heavy warships.

This is because 3/4 of cargo space is potentially outfit space, and a freighter with decent cargo space is potentially an OP warship. We likewise have warships with none/terrible cargo space. A hard cap on outfit space is needed, so that the two don't have to be balanced against each other.

I know it seems like a hard sell, but you really need to consider the variety it opens up in stock ship options. Even a Worldship could regain its cargo space. Freighters can not suck.

**Comment by Architect-Luke** [comment]
The elephant in the room is that it fixes the Bactrian for tiers over 1 without giving it a significant performance nerf in its own tier. And less directly, it ends the flagship meta when warships etc can have better cargo space.

**Comment by Nechochwen-D** [comment]
@Architect-Luke While I tend to agree with you that this issue needs to be addressed, the cargo expansions directly modify the ship stats while the ramscoop value is actually run through a section of the Ship::Move function where it is possible to apply more advanced math.

**Comment by MessyMix** [comment]
>Right now we have large freighters with less cargo space than some heavy warships.

This literally sounds like it's begging for the Bactrian nerf, and you manage to see it as a problem with Mass Expansions.

Take away the Bactrian, this problem doesn't exist.

I do see a possibility with diminishing returns, but then it will make it hard to min-max ships, which I know a lot of players (myself included) try to do.

**Comment by Wrzlprnft** [comment]
> The elephant in the room is that it fixes the Bactrian for tiers over 1 without giving it a significant performance nerf in its own tier. And less directly, it ends the flagship meta when warships etc can have better cargo space.

i am confused by this.
while mz's idea would directly hit the only ship one uses mass mass expansions on, your proposal, especially, basically implying the same, does not really... kinda... hit it.
of you are purely for removing mass expansions, the bactrian still has thr crew capacity, the hardpoints, the engine space and weapon space to be everything. it is true that one could not make it a super shield tank by stacking only shields, or would habe to cut on eventual skylances (cough). but otherwise it hits every single other ship in the game, but makes the bactrian even better. if i compare a war behemoth, which are still slow as hell, to it, the ship looses nothing with mass expansion, because it kinda sucks as a warship anyway. while actual warships would loose the ability to make use of their cargonspace as outfit space, or making a warship which a little more cargo space to plunder or make missions alongside other stuff.

**Comment by kikotheexile** [comment]
I actually really love MZ's idea when it comes to mass expansions messing up structural integrity and decreasing heat dissipation, I mean, you're literally knocking down and putting up walls and structure. Do the same to a house and it loses structural integrity, too. Thermodynamics teaches us that the more solid surface area you have, the more heat that structure can absorb and dissipate. Further, messing with electronics and electrical conduits would modulate the shield shape by cutting off some shield emitters from their generator source, thus weakening the shield strength overall. 

That said, this is the answer to the wrong question. This issue and the issue with the Bactrian are almost entirely unrelated.

**Comment by Architect-Luke** [comment]
If you don't see how a Bactrian that can't carry triple cores and shield generators around with their weapons is not nerfed, then I won't continue that line of discussion.

This seems to only hit Bactrian because every

**Comment by Architect-Luke** [comment]
Gosh this sucks as a forum. This seems to only hit the Bactrian because you are not acknowledging that cargo space on every warship in the game is made intentionally terrible so that it won't become another Bactrian, while cargo ships themselves often have bad cargo space and always bad stats for the same reason.

**Comment by Wrzlprnft** [comment]
i am aware that it nerfes the bactrian. though, it nerfes everything else accordingly, so the bactrian keeps it's position. maybe except when we move to higher tiers, but that point would come nonetheless, even with the current system.

**Comment by kikotheexile** [comment]
It literally nerfs everything, causing zero change overall to the baccy itself...

**Comment by kikotheexile** [comment]
Wrzl beat me to it.

**Comment by Wrzlprnft** [comment]
hehe.

don't get me wrong, luke, “i see what you did there“. it just not solves the problem, in my opinion.

**Comment by kikotheexile** [comment]
>Literally one ship is the problem, throwing the balance off and kinda going against its own lore
"Lets change every ship in the game in order to avoid fixing this one ship because reasons."
Ad nauseam. 

The mass expansions being revisited is a good idea for reasons other than the baccy, and revisiting them would not fix the baccy.

**Comment by Fleurghber** [comment]
I had some thoughts about this a while ago: I don’t know how much adjustment of the game would be required (apart from creating some new categories and adjusting the ship stats to work under the new system), but if it’s feasible, how about something like this?

Both weapon and engine space require an equivalent amount of free outfit space to be used.  Apply the same restrictions to cargo (and cabin) space?

There would then be opportunities for up to four types of conversion:

Outfit <-> Cargo

Outfit <-> Engine

Outfit <-> Weapon

Outfit <-> Cabin

In ship terms, it might not amount to anything more than moving a wall or deck zone marker, but having to be carried out by an expert to ensure a satisfactory job, hence the fee.  And like Escape Velocity, there could be the opportunity for a retool, making more efficient use of space but non-reversible.

A further restriction: to maintain balance, it should not be possible for the total amount of unused cabin, cargo, engine and weapon space to be greater than the amount of unused outfit space.

Would this be a sensible solution, or am I just muddying the waters further?

**Comment by Amazinite** [comment]
Engine/Weapon <-> Outfit doesn't really make sense. All the outfit space of your ship is where outfits are placed, but some of that space is specialized to engines and weapons. It's like having a room where you store stuff (your outfit space). Let's say a quarter of the room is specialized to storing frozen goods (let's say weapon space), so you can only store frozen goods (weapons) in that part of the room. Now you could still store other stuff in that part of the room (other outfits like shields), but you can't store frozen goods anywhere else in the room (only so much of your outfit space can be weapon space). You couldn't "convert" that room to lessen how many frozen goods you can store while increasing your overall storage space. It's like getting something out of nothing. In short, this doesn't make sense because weapons and engines take up both outfit space and weapon/engine space. If they only took weapon space or only took engine space, then it would make more sense. 

Also, Outfit <-> "Cabin" has been noed. Outfit -> Cabin is already a thing with bunk rooms, but Outfit <- Cabin has been shot down on the grounds that it will make things even worse than mass expansions, because high-bunk ships would end up having even more outfit space stacked on top of what they can already convert from cargo.

**Comment by Architect-Luke** [comment]
Well, any more ideas on balancing mass expansions? I really believe that they currently limit our ship design pretty severely.  I think decreasing returns is an excellent way to do it.  I would love to see cargo capable ships that are not so severely gimped.

**Comment by Fleurghber** [comment]
I agree with the points raised, but some balance could be retained by bigger/rarer/more complex/powerful items requiring more specialist crew members to use them, as is already the case with some weapons, so scrapping too many bunks wouldn't work.

**Comment by Architect-Luke** [comment]
@Fleurghber that sounds like a headache that would reduce all customization severely. It would have to, because the penalties would need to be significant for it to work. Players would have a hard time even figuring out whether they could do the modifications they want, and they would have very limited options on most ships, unless you upped bunk capacity... which would undo your "balance."

Anyway, I think that's functionally off-topic.

**Comment by jafdy** [comment]
>reduce all customization severely

That's what you've been going on about doing this whole time.

**Comment by endless-sky** [comment]
Well, I think I've got a solution.

Adding attributes to the ships themselves (say, by having each chassis have an "expansions" attribute that each mass expansion consumes) is a non-starter because ships that players bought before the change would not have that attribute and thus could not install mass expansions.

So, the Mass Expansion itself needs attributes that makes 3-4 of them usable, but add up to an impossible penalty once you have more than about 10 of them. None of the existing ship attributes serve that purpose well enough; the best I found was a combination of drag and reduced shields and hull, but that feels rather arbitrary.

So, I'm thinking of adding a new attribute, "cooling inefficiency." The more of it you have, the less benefit you get from your cooling outfits, in an S-curve:

![efficiency curve](https://cloud.githubusercontent.com/assets/11450262/24386174/25637678-133c-11e7-9d79-b06665d585bb.png)

For example, if you have 4 mass expansions, a large heat shunt will only give you 2722 cooling (3240 * 84%). If you have 10, the shunt only gives 940 cooling. Eventually you reach a point where all the space you gain with the expansion would have to be filled with cooling just to break even.

The elegance of this is that it lets you expand a ship considerably as long as you're running cool outfits, but if you want to use any of the most powerful engines or weapons or generators you need to stick to about 10 expansions on a Bactrian instead of 26. That creates a new reason to use high-efficiency outfits like solar collectors. And, it lets you still turn a Bactrian into a transport, but only if you're okay with it being useless in battle.

**Comment by MessyMix** [comment]
Wow.

Two birds with one stone? #2076 is also addressed in this solution.

**Comment by Wrzlprnft** [comment]
i like it, too. though, that needs to be introduced to the players as well.

**Comment by Bladewood** [comment]
A clever idea.

How do we lore this out? I don't really see how mass expansions correlate to efficient cooling.

**Comment by Amazinite** [comment]
Love it.

**Comment by LocalGod79** [comment]
@Bladewood "...since cargo space is seldom equipped to deal with the heat loads of high-tech outfits, heat dissipation on converted space is far less efficient than stock outfit space..."

higher end cooling technologies may push the "combat-usable mass-mass expansion limit" beyond 10, but probably not *much* higher, since the losses are exponential, and hopefully it makes more sense to use one of the ships from those higher tiers than make 15 or 30 more outfit space...

**Comment by endless-sky** [comment]
The idea is that when ship designers determine the layout for a new model
of ship, they carefully account for the amount of heat that the outfit
sections will produce and either lay them out in a way that avoids hot
spots, or add built-in heating ducts or heat sinks to wick the excess heat
out to the hull. (Basically, the same way that CPU designers spread out the
components of the chip that draw the most power.)

If you start rearranging the interior of your ship, that careful balance
gets out of whack, and it also becomes harder to lay out extra cooling
ducts and stuff like that.
